platform-sdk/docs/proposals/healthMonitor/healthMonitorProposal.md

```diff
@@ -0,0 +1,127 @@
+it reports the amount of time that the most unhealthy queue has been unhealthy. If there is one queue that has been
+unhealthy for 10s and another that has been unhealthy for 100s, it will report that the system unhealthiness factor
+is 100s. If the second queue clears its backlog and becomes healthy, it will then report that the system unhealthiness
+factor is 10s.
+
+The health monitor publishes health reports over a wire. If it sends a duration of 0s, that means that the system is in
+a healthy state. As the system becomes more and more unhealthy, it will periodically publish higher and higher
+durations.
+
+#### Reaction
+
+Various subsystems react to the health monitor's reports in different ways.
+
+##### Event Creation
+
+If the health monitor reports an unhealthy duration exceeding X seconds (configurable), then the creation of self
+events is disabled. Once the system reports an unhealthy duration below X seconds, event creation may resume.
+
+##### Gossip
+
+The sync gossip algorithm has the concept of a "sync permit". In order to initiate a sync operation, a permit must
+be acquired. When the sync is completed, the permit is released. If there are 5 permits available (configurable),
+then at any point in time there may be at most 5 sync operations in progress.
+
+When the health monitor reports that the system is becoming healthy, the gossip system takes no action until the
+unhealthy duration exceeds the grace period (a configurable duration). This is designed to prevent the gossip
+subsystem from reacting too aggressively to little blips.
+
+When the system is unhealthy for a duration that exceeds the grace period, the gossip system will begin revoking
+permits. Revoking a permit does not halt a sync operation that is currently in progress, but it can prevent a new
+sync operation from starting. For each second that the system is unhealthy beyond the grace period, permits are revoked
+at the rate specified by configuration.
+
+When the system eventually becomes healthy again, permits are gradually returned according to the rate specified in
+configuration.
+
+There is a special edge case when permits are being returned. Suppose the system has been unhealthy for a while, and
+then suddenly the backlog of work is cleared and the system becomes healthy. If all permits have been revoked, then
+it may take several seconds for the first permit to be returned. This means that a healthy system may have zero permits
+for a time after becoming healthy. To combat this problem, there is a configurable "minimum healthy permits" value.
+When the system is healthy, this is the minimum number of un-revoked permits that will be available. Note that this
+value is not respected in an unhealthy system, which should eventually have 0 permits if the problem is not resolved.
+
+##### Acceptance of application transactions
+
+If the health monitor reports an unhealthy duration exceeding X seconds (configurable), then the acceptance of
```

Comment:
How does this fit in with platform status and CHECKING? If the monitor determines that a node is unhealthy, will it put the node into CHECKING? Can an unhealthy node initiate a reconnect or teach a reconnect?